extension/src/state_aggregate.rs

```diff
@@ -33,20 +33,59 @@ pub mod toolkit_experimental {
             states_len: u64, // TODO JOSH this and durations_len can be 32
             durations_len: u64,
             durations: [DurationInState; self.durations_len],
+            first_time: i64,
+            last_time: i64,
+            first_state: u32,
+            last_state: u32,  // first/last state are idx into durations, keep together for alignment
             states: [u8; self.states_len],
         }
     }
 
     impl StateAgg<'_> {
-        pub fn new(states: String, durations: Vec<DurationInState>) -> Self {
+        pub(super) fn new(
+            states: String,
+            durations: Vec<DurationInState>,
+            first: Option<Record>,
+            last: Option<Record>,
+        ) -> Self {
+            assert!(
+                (durations.is_empty() && first.is_none() && last.is_none())
+                    || (!durations.is_empty() && first.is_some() && last.is_some())
+            );
             let states_len = states.len() as u64;
             let durations_len = durations.len() as u64;
+            let mut first_state = durations.len();
+            let mut last_state = durations.len();
```

Comment:
Hmm, this was more about initializing them to values that would fail the assertion at line 77 if we didn't overwrite them inside the loop.  However, you are right that this becomes easier to follow if we handle the empty case separately at the top of this function.  The existing behavior allows for calling new with empty arguments, so I don't want to move that into a `Default` as part of this change.